### PR TITLE
Bugfix: Conflicting UUIDs

### PR DESCRIPTION
**Pull request by DarkWugWug**
There are test that are not necessary for the bug fix! If you don't want them, the only change necessary is in _src/client.rs_.

I wrote test for the entities structure as a jumping in point and to sanity check myself. I think that they are a value add so I'm leaving them in with this bug fix.

### CODE REVIEW

File: src/entity.rs [hunk 1]
```diff
@@ -648,3 +648,123 @@ impl From<EntitySpawnPacket> for S2cPlayPacket {
...
+    use uuid::Uuid;
+    use super::{Entities, EntityId, EntityKind};
+    use crate::{
+        config::Config,
+        slab_versioned::Key,
+        server::Server
+    };
+
+    /// Created for the sole purpose of use during unit tests.
+    pub struct MockConfig { }
```
Comment (rj00a) [inline src/entity.rs:664]:
```suggestion
    struct MockConfig;
```

**Review by rj00a — CHANGES_REQUESTED** [review]
Nice find.

File: src/config.rs [hunk 2]
```diff
@@ -8,6 +8,8 @@ use tokio::runtime::Handle as TokioHandle;
 
 use crate::biome::Biome;
 use crate::dimension::Dimension;
+use crate::entity::EntityId;
+use crate::player_list::PlayerListId;
```
[the remark this answers is not in the document]
Comment (DarkWugWug) [reply src/config.rs]: Removed them

**Comment by dyc3** [comment]
How do we know for sure that we aren't still vulnerable to race conditions here?

**Comment by DarkWugWug** [comment]
## Analysis of the bug

For `client.created_this_tick()` the internal code was set to clear it after every update:
https://github.com/valence-rs/valence/blob/96f561494190338ed0186561135e1ce14cb9a177/src/client.rs#L1384

And that's reasonable because after this the client would not of been _created this tick_ in the next update loop, but the library wasn't doing this if a client had disconnected:
https://github.com/valence-rs/valence/blob/96f561494190338ed0186561135e1ce14cb9a177/src/client.rs#L907-L910

## Okay but why does it fix the terrain example

1. After connecting to server a new `Client` is created
2. Terrain's `Game.update()` is called and sees the newly created client and goes on incrementing the player count and creating an new player entity for it.
     > At this point `client.is_disconnected()` and `client.is_created_this_tick()` must be their default values. So `false` for disconnect and `true` for created this tick. This is because these values are only ever modified in the `Client.update()` which is called after `Game.update()`.
3. The library's `Client.update()` is called and sees that the network connection has been disconnected and destroys the client. Calls to `client.is_disconnected()` are now `true`. Then this fix changes `client.created_this_tick()` to `false`. Finally, **early return**.
4. Terrain's `Game.update()` is called and sees a client that is disconnected and disconnects it. This is unlike before when it would still think that this client needed to join!
5. Happy code noises 🦦🦀👌

**Pull request merged**